src/daos-deserialize/daos-deserialize.c

```diff
@@ -0,0 +1,420 @@
+            (*paths)[i] = MFU_STRDUP(argpaths[i]);
+            if ((*paths)[i] == NULL) {
+                MFU_LOG(MFU_LOG_ERR, "Failed to allocate paths.");
+                rc = 1;
+                goto out;
+            }
+        }
+    }
+out:
+    if (rc != 0) {
+        if (*paths != NULL) {
+            for (i = 0; i < num_files; i++) {
+                mfu_free(&(*paths)[i]); // TODO check this pointer handling
+            }
+            mfu_free(paths);
+        }
+    }
+    return rc;
+}
+
+int main(int argc, char** argv)
+{
+    /* assume we'll exit with success */
+    int rc = 0;
+
+    /* initialize MPI */
+    MPI_Init(&argc, &argv);
+    mfu_init();
+
+    /* get our rank */
+    int rank;
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+    /* verbose by default */
+    mfu_debug_level = MFU_LOG_VERBOSE;
+
+    int option_index = 0;
+    static struct option long_options[] = {
+        {"pool"                 , required_argument , 0, 'p'},
+        {"verbose"              , no_argument       , 0, 'v'},
+        {"quiet"                , no_argument       , 0, 'q'},
+        {"help"                 , no_argument       , 0, 'h'},
+        {0                      , 0                 , 0,  0 }
+    };
+
+    /* Parse options */
+    unsigned long long bytes = 0;
+
+    /* DAOS vars */ 
+    daos_args_t* daos_args = daos_args_new();    
+
+    int usage = 0;
+    while(1) {
+        int c = getopt_long(
+                    argc, argv, "p:vqh",
+                    long_options, &option_index
+                );
+
+        if (c == -1) {
+            break;
+        }
+
+        switch(c) {
+            case 'p':
+                uuid_parse(optarg, daos_args->src_pool_uuid);
+                break;
+            case 'v':
+                mfu_debug_level = MFU_LOG_VERBOSE;
+                break;
+            case 'q':
+                mfu_debug_level = MFU_LOG_NONE;
+                break;
+            case 'h':
+                usage = 1;
+                break;
+            case '?':
+                usage = 1;
+                break;
+            default:
+                if(rank == 0) {
+                    printf("?? getopt returned character code 0%o ??\n", c);
+                }
+        }
+    }
+
+    /* If we need to print the usage
+     * then do so before internal processing */
+    if (usage) {
+        if (rank == 0) {
+            print_usage();
+        }
+        mfu_finalize();
+        MPI_Finalize();
+        return 1;
+    }
+
+    int tmp_rc;
+    char** argpaths = (&argv[optind]);
+
+    /* The remaining arguments are treated as src/dst paths */
+    int numpaths = argc - optind;
+
+    /* advance to next set of options */
+    optind += numpaths;
+
+    if (numpaths < 1) {
+        MFU_LOG(MFU_LOG_ERR, "At least one file or directory is required:"
+                MFU_ERRF, MFU_ERRP(-MFU_ERR_INVAL_ARG));
+        mfu_finalize();
+        MPI_Finalize();
+        return 1;
+    }
```

Comment:
@dsikich We also need to make sure the --pool is passed in. Not supplying one currently gives an error.

Reply:
> @dsikich If I try to serialize a container with more ranks than there are objects in the container, I get:
> 
> ```
> [/home/dbohninx/mpifileutils_dev/mpifileutils/src/daos-serialize/daos-serialize.c:202] ERROR: Failed to serialize container (0)
> ```
> 
> For EACH "extra" rank. And then the application hangs (maybe because of the previous reduction issue noted before).

@daltonbohning yeah, it is because not all ranks call Allreduce. I think I can find a way to move this out of the handler and just pass in the total to it.

Reply:
@daltonbohning I might also just be able to split MPI_COMM_WORLD and call Allreduce on only the ranks doing work, thanks